cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java

```diff
@@ -131,19 +130,23 @@ public boolean hasNamespace(Id.Namespace namespaceId) {
    * @param metadata the {@link NamespaceMeta} for the new namespace to be created
    * @throws AlreadyExistsException if the specified namespace already exists
    */
-  public void createNamespace(NamespaceMeta metadata) throws AlreadyExistsException, IOException {
+  public void createNamespace(NamespaceMeta metadata) throws NamespaceCannotBeCreatedException, AlreadyExistsException {
     // TODO: CDAP-1427 - This should be transactional, but we don't support transactions on files yet
-    NamespaceMeta existing = store.createNamespace(metadata);
+    Preconditions.checkArgument(metadata != null, "Namespace metadata should not be null.");
+    NamespaceMeta existing = store.getNamespace(Id.Namespace.from(metadata.getId()));
     if (existing != null) {
       throw new AlreadyExistsException(NAMESPACE_ELEMENT_TYPE, metadata.getId());
     }
-    Location namespaceLocation = locationFactory.create(metadata.getId());
-    String namespacePath = namespaceLocation.toURI().getPath();
-    if (namespaceLocation.exists()) {
-      throw new IOException(String.format("Namespace home directory %s already exists", namespacePath));
+
+    try {
+      dsFramework.createNamespace(Id.Namespace.from(metadata.getId()));
+    } catch (DatasetManagementException e) {
+      throw new NamespaceCannotBeCreatedException(metadata.getId(), e.getMessage());
     }
-    if (!namespaceLocation.mkdirs()) {
-      throw new IOException(String.format("Error while creating namespace home directory %s", namespacePath));
+
+    existing = store.createNamespace(metadata);
+    if (existing != null) {
```

Comment:
Yes agreed it's a little confusing. 

However, namespace create is not atomic, so added an extra check in the end to make sure if it exists, given that ``createNamespace`` already returns  the namespace if it exists.